src/EnrollWebauthnController.js

```diff
@@ -0,0 +1,157 @@
+/*!
+ * Copyright (c) 2015-2016, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+define([
+  'okta',
+  'util/FormType',
+  'util/FormController',
+  'util/CryptoUtil',
+  'views/enroll-factors/Footer',
+  'vendor/lib/q',
+  'views/mfa-verify/HtmlErrorMessageView'
+],
+function (Okta, FormType, FormController, CryptoUtil, Footer, Q, HtmlErrorMessageView) {
+
+  var _ = Okta._;
+
+  return FormController.extend({
+    className: 'enroll-u2f',
+    Model: {
+      local: {
+        '__enrolled__': 'boolean'
+      },
+
+      save: function () {
+        this.trigger('request');
+
+        if (this.get('__enrolled__')) {
+          return this.activate();
+        }
+
+        return this.doTransaction(function (transaction) {
+          var factor = _.findWhere(transaction.factors, {
+            factorType: 'webauthn',
+            provider: 'FIDO'
+          });
+          return factor.enroll();
+        });
+      },
+
+      activate: function () {
+        this.set('__enrolled__', true);
+        this.trigger('errors:clear');
+
+        return this.doTransaction(function (transaction) {
+          var activation = transaction.factor.activation;
+          var deferred = Q.defer();
```

Comment:
Nit: No need to have a separate defer for this. Since we're already creating a promise with `navigator.credentials.create`, we can just use that:

```js
return navigator.credentials.create({ publicKey: options })
  .then(function () {
    return transaction.activate(/* stuff */);
  })
  .catch(function () {
    throw new Error({ xhr: 'stuff' });
  });
```

Reply:
Also, no need for all the names of "publicKey" - we go from activation, to options, to publickey. Why not just:

```js
var publicKey = _.extend({}, transaction.factor.activation, {
  challenge: stuff,
  user: {
    id: stuff
  }
});
```